connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -1133,6 +1139,105 @@ public void setTargetState(String connName, TargetState state, Callback<TargetSt
+        try {
+            ListConsumerGroupOffsetsResult listConsumerGroupOffsetsResult = admin.listConsumerGroupOffsets(groupId);
+            try {
+                // Not using a timeout for the Future::get here because each offset get request is handled in its own thread in AbstractHerder
+                // and the REST API request timeout in HerderRequestHandler will ensure that the user request doesn't hang indefinitely
+                Map<TopicPartition, OffsetAndMetadata> offsets = listConsumerGroupOffsetsResult.all().get().get(groupId);
+                return SinkUtils.consumerGroupOffsetsToConnectorOffsets(offsets);
+            } catch (InterruptedException | ExecutionException e) {
+                throw new ConnectException("Failed to retrieve consumer group offsets for sink connector " + connName, e);
+            }
+        } finally {
+            Utils.closeQuietly(admin, "Offset fetch admin for sink connector " + connName);
+        }
+    }
+
+    /**
+     * Get the current offsets for a source connector.
+     * @param connName the name of the source connector whose offsets are to be retrieved
+     * @param connector the source connector
+     * @param connectorConfig the source connector's configurations
+     * @return the source connector's offsets
+     */
+    private ConnectorOffsets sourceConnectorOffsets(String connName, Connector connector, Map<String, String> connectorConfig) {
+        SourceConnectorConfig sourceConfig = new SourceConnectorConfig(plugins, connectorConfig, config.topicCreationEnable());
+        ConnectorOffsetBackingStore offsetStore = config.exactlyOnceSourceEnabled()
+                ? offsetStoreForExactlyOnceSourceConnector(sourceConfig, connName, connector)
+                : offsetStoreForRegularSourceConnector(sourceConfig, connName, connector);
+        CloseableOffsetStorageReader offsetReader = new OffsetStorageReaderImpl(offsetStore, connName, internalKeyConverter, internalValueConverter);
+        return sourceConnectorOffsets(connName, offsetStore, offsetReader);
+    }
+
+    // Visible for testing
+    ConnectorOffsets sourceConnectorOffsets(String connName, ConnectorOffsetBackingStore offsetStore,
+                                            CloseableOffsetStorageReader offsetReader) {
+        offsetStore.configure(config);
+        offsetStore.start();
+        try {
+            Set<Map<String, Object>> connectorPartitions = offsetStore.connectorPartitions(connName);
+            List<ConnectorOffset> connectorOffsets = offsetReader.offsets(connectorPartitions).entrySet().stream()
+                    .map(entry -> new ConnectorOffset(entry.getKey(), entry.getValue()))
+                    .collect(Collectors.toList());
+            return new ConnectorOffsets(connectorOffsets);
+        } finally {
+            Utils.closeQuietly(offsetReader, "Offset reader for connector " + connName);
+            offsetStore.stop();
```

Comment:
IMO this should also be wrapped in `Utils::closeQuietly`. If not, users may spam the read endpoint with manual retries and it could exacerbate a potential resource leak caused by the failure to close the store. If we have a valid result to provide to them and the only issue is a failure to shut down the store, best to give that valid result to the user and log a warning.